app/services/assignment_service.py

```diff
@@ -267,60 +290,77 @@ def _compute_master_notebook_name(assignment_name: str) -> str:
         return f"{ assignment_name }-prof.ipynb"
 
 class InstructorAssignmentService(AssignmentService):
-    def __init__(self, session: Session, instructor: InstructorModel, assignment: AssignmentModel):
+    def __init__(self, session: Session, instructor_model: InstructorModel, assignment_model: AssignmentModel, course_model: CourseModel):
         super().__init__(session)
-        self.instructor = instructor
-        self.assignment = assignment
+        self.instructor_model = instructor_model
+        self.assignment_model = assignment_model
+        self.course_model = course_model 
 
-    def get_is_available(self) -> bool:
-        if not self.assignment.is_created: return False
+    # The release date for a specific student, considering extra_time
+    def get_adjusted_available_date(self) -> datetime | None:
+        assignment_open_date = self.assignment_model.available_date or self.course_model.start_at
+        if assignment_open_date is None: return None
+        
+        return assignment_open_date
 
-        current_timestamp = self.session.scalar(func.current_timestamp())
-        return current_timestamp >= self.assignment.available_date
-    
-    def get_is_closed(self) -> bool:
-        if not self.assignment.is_created: return False
+    # The due date for a specific student, considering extra_time
+    def get_adjusted_due_date(self) -> datetime | None:
+        assignment_due_date = self.assignment_model.due_date or self.course_model.end_at
+        if assignment_due_date is None: return None
+
+        return assignment_due_date
+
+    def get_assignment_status(self) -> AssignmentStatus:
+        if not self.assignment_model.is_published: return AssignmentStatus.UNPUBLISHED
 
         current_timestamp = self.session.scalar(func.current_timestamp())
-        return current_timestamp > self.assignment.due_date
+        adjusted_available_date = self.get_adjusted_available_date()
+        adjusted_due_date = self.get_adjusted_due_date()
+        
+        if adjusted_available_date >= current_timestamp: return AssignmentStatus.UPCOMING
```

Comment:
this comparison will cause an error if `adjusted_available_date` is None. which is one of the possibilities of values that can be returned by `get_adjusted_available_date`

Reply:
Check this logic now, I think I added a sufficient null check

Reply:
The decision I made is that if an assignment doesn't have a available date _or_ a course open date set (which is how `adjusted_available_date` becomes None) then I'm putting the assignment in the UPCOMING status, and if it does have an available date of some form but doesn't have a due date _or_ a course end date set (which is how adjusted_due_date becomes None) then I mark the assignment with status CLOSED. That way it never accidentally becomes an assignment the student can work on when we haven't set up the course correctly yet. Fixing this is easy in the configuration: just set a course start and end date at a minimum.

Reply:
@frostyfan109 Tagging you too so you are bought in to this logic

Reply:
Yup I like it, not resolving the convo so that we dont lost this, but I am okay with this approach